akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -0,0 +1,170 @@
+          case None                         ⇒
+        }
+        unmarked -= u // permanent mark
+        tempMark -= u
+        result = u :: result
+      }
+    }
+
+    result.reverse
+  }
+
+}
+
+final class CoordinatedShutdown private[akka] (
+  system:                   ExtendedActorSystem,
+  private[akka] val phases: Map[String, CoordinatedShutdown.Phase]) extends Extension {
+
+  private val log = Logging(system, getClass)
+  private val knownPhases = phases.keySet ++ phases.values.flatMap(_.dependsOn)
+  private val orderedPhases = CoordinatedShutdown.topologicalSort(phases)
+  private val tasks = new ConcurrentHashMap[String, Vector[() ⇒ Future[Done]]]
+  private val runStarted = new AtomicBoolean(false)
+  private val runPromise = Promise[Done]()
+
+  /**
+   * Add a task to a phase. It doesn't remove previously added tasks.
+   * Tasks added to the same phase are executed in parallel without any
+   * ordering assumptions. Next phase will not start until all tasks of
+   * previous phase have been completed.
+   */
+  @tailrec def addTask(phase: String)(task: () ⇒ Future[Done]): Unit = {
+    require(
+      knownPhases(phase),
+      s"unknown phase [$phase], known phases [$knownPhases]. " +
+        "All phases (along with their optional dependencies) must be defined in configuration")
+    val current = tasks.get(phase)
+    if (current == null) {
+      if (tasks.putIfAbsent(phase, Vector(task)) != null)
+        addTask(phase)(task) // CAS failed, retry
+    } else {
+      if (!tasks.replace(phase, current, current :+ task))
+        addTask(phase)(task) // CAS failed, retry
+    }
+  }
+
+  def run(): Future[Done] = {
+    if (runStarted.compareAndSet(false, true)) {
+      import system.dispatcher
+      val debugEnabled = log.isDebugEnabled
+      def loop(remainingPhases: List[String]): Future[Done] = {
+        remainingPhases match {
+          case Nil ⇒ Future.successful(Done)
+          case phase :: remaining ⇒
+            (tasks.get(phase) match {
+              case null ⇒
+                if (debugEnabled) log.debug("Performing phase [{}] with [0] tasks", phase)
+                Future.successful(Done)
+              case tasks ⇒
+                if (debugEnabled) log.debug("Performing phase [{}] with [{}] tasks", phase, tasks.size)
+                // note that tasks within same phase are performed in parallel
+                val result = Future.sequence(tasks.map { task ⇒
+                  try task.apply() catch { case NonFatal(e) ⇒ Future.failed(e) }
+                }).map(_ ⇒ Done)
+                val timeout = phases(phase).timeout
+                val timeoutFut = after(timeout, system.scheduler)(Future.failed(
+                  new TimeoutException(s"Coordinated shutdown phase [$phase] timed out after $timeout")))
+                val resultWithTimeout = Future.firstCompletedOf(List(result, timeoutFut))
+                if (phases(phase).recover)
+                  resultWithTimeout.recover {
```

Comment:
this is wrong, recover should be done on the individual task, otherwise next phase may start too early